File: app/src/main/java/com/cliniva/enventory/ui/inventory/InventoryFragment.java
package com.cliniva.enventory.ui.inventory;


import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.cliniva.enventory.R;
import com.cliniva.enventory.adapter.RecyclerViewAdapter;
import com.cliniva.enventory.adapter.base.BaseRecyclerClickListener;
import com.cliniva.enventory.adapter.base.BaseRecyclerViewHolder;
import com.cliniva.enventory.event.OnProductItemClickedListener;
import com.cliniva.enventory.model.Brand;
import com.cliniva.enventory.model.Product;
import com.cliniva.enventory.ui.base.BaseFragment;
import com.cliniva.enventory.viewholder.BrandHolder;
import com.cliniva.enventory.viewholder.ProductHolder;

import java.util.List;

/**
 * A simple {@link Fragment} subclass.
 */
public class InventoryFragment extends BaseFragment implements InventoryContract.View, BaseRecyclerClickListener<Brand> {

    private InventoryContract.Presenter mInventoryPresenter;
    private RecyclerView mProductList;
    private RecyclerView mBrandList;

    public InventoryFragment() {
        // Required empty public constructor
    }

    public static InventoryFragment getInstance(){
        InventoryFragment fragment = new InventoryFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public int getLayoutRes() {
        return R.layout.fragment_inventory;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mInventoryPresenter = new InventoryPresenter(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        setToolbar(view, mInventoryPresenter.getContext().getString(R.string.action_inventory), false);

        mProductList = view.findViewById(R.id.rv_product_list);
        mBrandList = view.findViewById(R.id.rv_brand_name_list);
        LinearLayoutManager verticalManager = new LinearLayoutManager(mInventoryPresenter.getContext());
        mProductList.setLayoutManager(verticalManager);
        LinearLayoutManager horizontalManager = new LinearLayoutManager(mInventoryPresenter.getContext(), LinearLayoutManager.HORIZONTAL, false);
        mBrandList.setLayoutManager(horizontalManager);
        mInventoryPresenter.onLoadList();
    }

    @Override
    public void setBrandListToView(List<Brand> brandList) {
        RecyclerViewAdapter<Brand, BaseRecyclerClickListener<Brand>> mBrandAdapter = new RecyclerViewAdapter<Brand, BaseRecyclerClickListener<Brand>>(brandList) {
            @NonNull
            @Override
            public BaseRecyclerViewHolder<Brand, BaseRecyclerClickListener<Brand>> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.item_brand, parent, false);
                return new BrandHolder(binding);
            }
        };
        mBrandAdapter.setListener(this);
        mBrandList.setAdapter(mBrandAdapter);
    }

    @Override
    public void setProductListToView(List<Product> productList) {
        RecyclerViewAdapter<Product, OnProductItemClickedListener> mProductAdapter = new RecyclerViewAdapter<Product, OnProductItemClickedListener>(productList) {
            @NonNull
            @Override
            public BaseRecyclerViewHolder<Product, OnProductItemClickedListener> onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.item_inventory, parent, false);
                return new ProductHolder(binding);
            }
        };

        mProductAdapter.setListener(new OnProductItemClickedListener() {
            @Override
            public void onSingleViewClicked(View v, int position) {

            }

            @Override
            public void onItemClickListener(Product item, int position) {

            }
        });
        mProductList.setAdapter(mProductAdapter);
    }

    @Override
    public void onItemClickListener(Brand item, int position) {

    }
}

File: app/src/main/java/com/amirsons/inventory/ui/fragment/inventory/InventoryFragment.kt
package com.amirsons.inventory.ui.fragment.inventory

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearSnapHelper
import com.amirsons.inventory.R
import com.amirsons.inventory.datamanager.model.BrandProduct
import com.amirsons.inventory.datamanager.model.Product
import com.amirsons.inventory.event.OnProductItemClickedListener
import com.amirsons.inventory.recyclerview.base.BaseRecyclerClickListener
import com.amirsons.inventory.recyclerview.base.BaseRecyclerViewHolder
import com.amirsons.inventory.recyclerview.base.RecyclerViewAdapter
import com.amirsons.inventory.recyclerview.viewholder.BrandHolder
import com.amirsons.inventory.recyclerview.viewholder.ProductHolder
import com.amirsons.inventory.ui.base.BaseFragment
import com.amirsons.inventory.ui.widgets.BottomSheetAddProduct
import com.amirsons.inventory.ui.widgets.OnAddProductConfirm
import kotlinx.android.synthetic.main.fragment_inventory.*

/**
 * A simple [BaseFragment] subclass.
 */
class InventoryFragment : BaseFragment(), InventoryView {

    private lateinit var mProductAdapter: RecyclerViewAdapter<Product, OnProductItemClickedListener>
    private lateinit var mBrandAdapter: RecyclerViewAdapter<BrandProduct, BaseRecyclerClickListener<BrandProduct>>
    private lateinit var mInventoryPresenter: InventoryPresenter

    private var defaultBrandPosition = -1

    override val contentLayout: Int
        get() = R.layout.fragment_inventory

    companion object {

        val instance: InventoryFragment
            get() {
                val fragment = InventoryFragment()
                val args = Bundle()
                fragment.arguments = args
                return fragment
            }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mInventoryPresenter = InventoryMvp(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // set toolbar name
        setToolbar(view, mInventoryPresenter.context.getString(R.string.action_inventory), false)

        // init recycler view
        initBrandRecyclerView()
        initProductRecyclerView()

        // load the product list
        mInventoryPresenter.onLoadProductList()

        // fab button for add product
        btn_fab.setOnClickListener {
            mInventoryPresenter.onAddProductClick()
        }
    }

    private fun initBrandRecyclerView() {

        val snapHelper = LinearSnapHelper()
        snapHelper.attachToRecyclerView(rv_brand_name_list)

        mBrandAdapter = object : RecyclerViewAdapter<BrandProduct, BaseRecyclerClickListener<BrandProduct>>() {
            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseRecyclerViewHolder<BrandProduct, BaseRecyclerClickListener<BrandProduct>> {
                return BrandHolder(inflate(R.layout.item_brand, parent))
            }
        }
        rv_brand_name_list.adapter = mBrandAdapter

        mBrandAdapter.setListener(object : BaseRecyclerClickListener<BrandProduct> {

            override fun onItemClickListener(item: BrandProduct, position: Int) {

                if (defaultBrandPosition != -1 && position != defaultBrandPosition){
                    mBrandAdapter.getItem(defaultBrandPosition).isSelected = false
                    mBrandAdapter.notifyItemChanged(defaultBrandPosition)
                }

                // set the product list
                mProductAdapter.setItems(item.productList)

                // store current selected position
                defaultBrandPosition = position

                rv_brand_name_list.scrollToPosition(position)
            }
        })
    }
    private fun initProductRecyclerView() {

        mProductAdapter = object : RecyclerViewAdapter<Product, OnProductItemClickedListener>() {
            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseRecyclerViewHolder<Product, OnProductItemClickedListener> {
                return ProductHolder(inflate(R.layout.item_inventory, parent))
            }
        }
        rv_product_list.adapter = mProductAdapter
    }

    override fun setBrandListToView(brandList: ArrayList<BrandProduct>) {

        mBrandAdapter.setItems(brandList)

        // set default product list to first item
        if (defaultBrandPosition == -1) {
            defaultBrandPosition = 0
        }

        brandList[defaultBrandPosition].isSelected = true
        mProductAdapter.setItems(brandList[defaultBrandPosition].productList)
    }


    override fun showAddProductBottomSheet(brandNameList: ArrayList<String>) {

        // show add product bottom sheet
        val addProductBottomSheet = BottomSheetAddProduct(context, brandNameList, object : OnAddProductConfirm {

            override fun onConfirm(product: Product): Boolean {

                mInventoryPresenter.onAddProduct(product)
                return true
            }
        })

        // show the bottom sheet dialog
        addProductBottomSheet.show()
    }
}
